extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js

```diff
@@ -65,17 +69,117 @@ describe('amp-ad-api-handler', () => {
+                  type: 'no-content',
+                }).then(() => {
+                  expect(noContentCallbackSpy).to.not.been.called;
+                });
+              });
+            });
       });
-      return startUpPromise.then(() => {
-        expect(iframe.style.visibility).to.equal('');
+
+      it('should resolve on message "no-content" if render-start is'
+          + 'implemented by 3P', () => {
+        adImpl.adType = 'doubleclick';
+        apiHandler = new AmpAdApiHandler(adImpl, adImpl.element, () => {});
+        const beforeAttachedToDom = element => {
+          element.setAttribute('data-amp-3p-sentinel', 'amp3ptest' + testIndex);
+          startUpPromise = apiHandler.startUp(element, true);
+        };
+        const noContentCallbackSpy
+            = sandbox.spy(apiHandler, 'noContentCallback_');
+        return createIframeWithMessageStub(window, beforeAttachedToDom)
+            .then(newIframe => {
+              iframe = newIframe;
+              expect(iframe.style.visibility).to.equal('hidden');
+              iframe.postMessageToParent({
+                sentinel: 'amp3ptest' + testIndex,
+                type: 'no-content',
+              });
+              return startUpPromise.then(() => {
+                expect(iframe.style.visibility).to.equal('');
+                expect(noContentCallbackSpy).to.be.calledOnce;
+              });
+            });
+      });
+
+      it('should NOT resolve on message "bootstrap-loaded" if render-start is'
+          + 'implemented by 3P', () => {
+        adImpl.adType = 'doubleclick';
+        apiHandler = new AmpAdApiHandler(adImpl, adImpl.element);
+        const beforeAttachedToDom = element => {
+          element.setAttribute('data-amp-3p-sentinel', 'amp3ptest' + testIndex);
+          startUpPromise = apiHandler.startUp(element, true);
+        };
+        return createIframeWithMessageStub(window, beforeAttachedToDom)
+            .then(newIframe => {
+              iframe = newIframe;
+              expect(iframe.style.visibility).to.equal('hidden');
+              iframe.postMessageToParent({
+                sentinel: 'amp3ptest' + testIndex,
+                type: 'bootstrap-loaded',
+              });
+            }).then(() => {
+              return expectPostMessage(iframe.contentWindow, window, {
+                sentinel: 'amp3ptest' + testIndex,
+                type: 'bootstrap-loaded',
+              }).then(() => {
+                const clock = sandbox.useFakeTimers();
+                clock.tick(0);
+                const timeoutPromise =
+                    timerFor(window).timeoutPromise(2000, startUpPromise);
+                clock.tick(2001);
+                return timeoutPromise.then(() => {
+                  throw Error('startUp resolve on bootstrap-loaded when'
+                      + 'when render-start is supported');
+                }, error => {
+                  expect(error).to.match(/timeout/);
+                });
```

Comment:
hmm, i saw usages of `.to.eventually.be.rejectedWith` already in the code base though.